ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/util/ConcurrencyLimiter.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.util.KsqlException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ConcurrencyLimiter {
```

Comment:
isn't this just a semaphore? I feel like we should use `java.util.concurrent.Semaphore` with `tryAcquire`  semantics to avoid some nasty unexpected concurrency bugs

Reply:
It is, though it returns a decrementer that ensures it happens just once.  There can be multiple code paths that try to decrement it, depending on if an exception happens and if pull queries were already launched in the background.  I added some comments to clarify this.

I ended up wrapping a semaphore so that we can choose to make it blocking in the future, if we want.